### PR TITLE
Fix n+1 problems

### PR DESCRIPTION
**Pull request by samuelmoiwa**
## **In this Branch, @KareemWilson  and I did the following:** :arrow_down: 

**1. Solve the N+1 problem when fetching all posts and their comments for a user.**

_Before Fixing this problem Here is the screenshot of the logs:_ 

Capture1

_After Fixing this problem Here is the screenshot of the logs:_ 

Capture2

**2. Made some Integration specs**

**3. Use Capybara to write integration tests for each view in your project.**: 

**4. Test the following actions:**

**_User index page:_**

- Show the username of all other users. :heavy_check_mark: 
- Show the profile picture for each user. :heavy_check_mark: 
- Show the number of posts each user has written. :heavy_check_mark: 
- When I click on a user, I am redirected to that user's show page. :heavy_check_mark: 

**_user show page:_**

- Show the user's profile picture. :heavy_check_mark: 
- Show the user's username. :heavy_check_mark: 
- Show the number of posts the user has written. :heavy_check_mark: 
- Show the user's bio. :heavy_check_mark: 
- Show the user's first 3 posts. :heavy_check_mark: 
- I can see a button that lets me view all of a user's posts. :heavy_check_mark: 
- When I click a user's post, it redirects me to that post's show page. :heavy_check_mark: 
- When I click to see all posts, it redirects me to the user's post's index page. :heavy_check_mark: 

**_User post index page:_**

- Show the user's profile picture. :heavy_check_mark: 
- Show the user's username. :heavy_check_mark: 
- Show the number of posts the user has written. :heavy_check_mark: 
- Show a post's title. :heavy_check_mark: 
- Show some of the post's body. :heavy_check_mark: 
- Show the first comments on a post.  :heavy_check_mark: 
- Show how many comments a post has. :heavy_check_mark: 
- Show how many likes a post has. :heavy_check_mark: 
- Show a section for pagination if there are more posts than fit on the view. :heavy_check_mark: 
- When I click on a post, it redirects me to that post's show page. :heavy_check_mark: 

**_Post show page:_**

- Show the post's title. :heavy_check_mark: 
- Show who wrote the post. :heavy_check_mark: 
- Show how many comments it has. :heavy_check_mark: 
- Show how many likes it has. :heavy_check_mark: 
- Show the post body. :heavy_check_mark: 
- Show the username of each commenter. :heavy_check_mark: 
- Show the comment each commenter left. :heavy_check_mark:

### CODE REVIEW

File: spec/views/post_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      expect(page).to have_content(@comment.text)
+    end
+
+    it 'should show pagination' do
+      expect(page).to have_content('Pagination')
+    end
+    it 'profile photo should be present' do
+      visit user_posts_path(@user.id)
+      expect(page).to have_xpath("//img[@src='#{@user.photo}']")
+    end
```
Comment (alexander16108) [inline spec/views/post_index_spec.rb:44]: Hi there, the following test required test case are not present in your post index spec file for testing please add them, make sure they pass and submit them for review again 😄 👍🏾 

- Tests that asserts you can see the first comments on a post.
- Tests that asserts you can see how many likes a post has.
- Tests that asserts you can see a section for pagination if there are more posts than fit on the view.

File: spec/views/post_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    it "should display the post's body" do
+      expect(page).to have_content(@post.text)
+    end
+
+    it 'should display the comments pusblished' do
+      @post.comments.each do |comment|
+        expect(page).to have_content(comment.text)
+        expect(page).to have_content(comment.user.name)
+      end
+    end
```
Comment (alexander16108) [inline spec/views/post_show_spec.rb]: Hi there, the following test required test cases are not present in your post index spec file for testing please add them, make sure they pass and submit them for review again 😄 👍🏾

- Tests that asserts you can see the username of each commentor.

File: spec/views/user_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      visit root_path
+      expect(page).to have_content(@user.name)
+      expect(page).to have_content(@user.posts_counter)
+      expect(page).to have_xpath("//img[@src='#{@user.photo}']")
+    end
+    it 'redirect to show' do
+      visit root_path
+      click_link @user.name, match: :first
+      expect(page).to have_current_path(root_path(@user))
+    end
```
Comment (alexander16108) [inline spec/views/user_index_spec.rb]: Hi there, the following test required test cases are not present in your post index spec file for testing please add them, make sure they pass and submit them for review again 😄 👍🏾

* Tests that assert you can see the username of all other users.
* Tests that assert you can see the profile picture for each user.

File: spec/views/user_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      click_link @posts.id, match: :first
+      expect(page).to have_current_path("/users/#{@user.id}/posts/#{@posts.id}")
+    end
+    it 'onclick redirect to posts index' do
+      visit user_path(@user)
+      click_button 'See all posts'
+      expect(page).to have_current_path("/users/#{@user.id}")
+    end
+  end
+  # rubocop:enable Layout/LineLength
```
Comment (alexander16108) [inline spec/views/user_show_spec.rb]: Hi there, the following test required test cases are not present in your post index spec file for testing please add them, make sure they pass and submit them for review again 😄 👍🏾

* Tests that assert you can see the user's profile picture.
* Tests that assert you can see the user's username.
* Tests that assert you can see the number of posts the user has written.

**Review by alexander16108 — CHANGES_REQUESTED** [review]
Hi @samuelmoiwa @KareemWilson , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights

- All linters are passing ✔️
- Descriptive commit messages ✔️
- Professional looking pull request ✔️

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by cynthiainga — APPROVED** [review]
Hi @samuelmoiwa, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights

- [ ] No linters errors ✔️ 
- [ ] Descriptive PR ✔️ 
- [ ] Correct Gitflow ✔️ 

## Optional suggestions

- [OPTIONAL] I noticed that your commit messages are not capitalized, the commit message should be capitalized. [Read more about how to write descriptive commit messages](https://www.freecodecamp.org/news/level-you-up-to-awesome-commit-messages-a85558cb90e8). Dont't forget this for your next commit 🙃 😉 

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**